File: js/src/progress.ts
import { native, NativeBox } from "./native";

export class ProgressToken {
  /** @internal */
  readonly _box: NativeBox;

  /** @internal */
  constructor(box_: NativeBox) {
    this._box = box_;
  }

  /** Create a root progress token at 0%. */
  static create(): ProgressToken {
    return new ProgressToken(native.progressNew());
  }

  /** Set this token's progress fraction (clamped to [0, 1]). */
  set(fraction: number): void {
    native.progressSet(this._box, fraction);
  }

  /** This token's progress fraction in [0, 1]. */
  get fraction(): number {
    return native.progressFraction(this._box);
  }

  /** This token's width as a fraction of the root [0, 1] range. */
  get width(): number {
    return native.progressWidth(this._box);
  }

  /** Whether this token's progress >= 1.0. */
  get isComplete(): boolean {
    return native.progressIsComplete(this._box);
  }

  /** Overall progress across the entire tree. */
  get rootFraction(): number {
    return native.progressRootFraction(this._box);
  }

  /** Split into subtokens by relative weights. */
  split(weights: number[]): ProgressToken[] {
    const boxes = native.progressSplit(this._box, weights);
    return boxes.map((b) => new ProgressToken(b));
  }

  /** Create one child subtoken covering `fracWidth` of this token's range. */
  subtoken(fracWidth: number): ProgressToken {
    return new ProgressToken(native.progressSubtoken(this._box, fracWidth));
  }

  clone(): ProgressToken {
    return new ProgressToken(native.progressClone(this._box));
  }
}

File: ts/cognee-ts-neon/src/progress.rs
use neon::prelude::*;

use cognee_core::ProgressToken;

use crate::error::throw_core_error;

pub struct NeonProgressToken {
    pub inner: ProgressToken,
}

impl Finalize for NeonProgressToken {}

pub fn progress_new(mut cx: FunctionContext) -> JsResult<JsBox<NeonProgressToken>> {
    Ok(cx.boxed(NeonProgressToken {
        inner: ProgressToken::new(),
    }))
}

pub fn progress_set(mut cx: FunctionContext) -> JsResult<JsUndefined> {
    let token = cx.argument::<JsBox<NeonProgressToken>>(0)?;
    let fraction = cx.argument::<JsNumber>(1)?.value(&mut cx);
    token.inner.set(fraction);
    Ok(cx.undefined())
}

pub fn progress_fraction(mut cx: FunctionContext) -> JsResult<JsNumber> {
    let token = cx.argument::<JsBox<NeonProgressToken>>(0)?;
    Ok(cx.number(token.inner.fraction()))
}

pub fn progress_width(mut cx: FunctionContext) -> JsResult<JsNumber> {
    let token = cx.argument::<JsBox<NeonProgressToken>>(0)?;
    Ok(cx.number(token.inner.width()))
}

pub fn progress_is_complete(mut cx: FunctionContext) -> JsResult<JsBoolean> {
    let token = cx.argument::<JsBox<NeonProgressToken>>(0)?;
    Ok(cx.boolean(token.inner.is_complete()))
}

pub fn progress_root_fraction(mut cx: FunctionContext) -> JsResult<JsNumber> {
    let token = cx.argument::<JsBox<NeonProgressToken>>(0)?;
    Ok(cx.number(token.inner.root_fraction()))
}

/// Split into subtokens by relative weights.
///
/// `progressSplit(token, weights: number[]) -> ProgressToken[]`
pub fn progress_split(mut cx: FunctionContext) -> JsResult<JsArray> {
    let token = cx.argument::<JsBox<NeonProgressToken>>(0)?;
    let weights_arr = cx.argument::<JsArray>(1)?;
    let len = weights_arr.len(&mut cx);
    let mut weights = Vec::with_capacity(len as usize);
    for i in 0..len {
        let w = weights_arr
            .get::<JsNumber, _, _>(&mut cx, i)?
            .value(&mut cx) as u32;
        weights.push(w);
    }

    let subtokens = token
        .inner
        .split(&weights)
        .or_else(|e| throw_core_error(&mut cx, e))?;

    let result = JsArray::new(&mut cx, subtokens.len());
    for (i, sub) in subtokens.into_iter().enumerate() {
        let js_sub = cx.boxed(NeonProgressToken { inner: sub });
        result.set(&mut cx, i as u32, js_sub)?;
    }
    Ok(result)
}

/// Create one child subtoken covering `fracWidth` of this token's range.
///
/// `progressSubtoken(token, fracWidth: number) -> ProgressToken`
pub fn progress_subtoken(mut cx: FunctionContext) -> JsResult<JsBox<NeonProgressToken>> {
    let token = cx.argument::<JsBox<NeonProgressToken>>(0)?;
    let frac_width = cx.argument::<JsNumber>(1)?.value(&mut cx);
    let sub = token.inner.subtoken(frac_width);
    Ok(cx.boxed(NeonProgressToken { inner: sub }))
}

pub fn progress_clone(mut cx: FunctionContext) -> JsResult<JsBox<NeonProgressToken>> {
    let token = cx.argument::<JsBox<NeonProgressToken>>(0)?;
    Ok(cx.boxed(NeonProgressToken {
        inner: token.inner.clone(),
    }))
}
